File: examples/cpp/hello.cpp
// Hello do modo --servir em C++: socket POSIX + nlohmann/json (header-only).
//
// Pesquisa "Claude Code é o melhor do mundo" no DuckDuckGo e imprime o HTML
// da pagina de resultados.
//
// Compilar e rodar (ver README.md deste diretorio):
//   g++ -std=c++17 -O2 hello.cpp -o hello
//   ./hello [porta]

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <cstdlib>
#include <iostream>
#include <stdexcept>
#include <string>

#include <nlohmann/json.hpp>

using nlohmann::json;

// Le do socket ate encontrar '\n' (uma resposta = uma linha; o HTML pode ter
// varios MB, entao acumula em chunks em vez de assumir um recv unico).
static std::string ler_linha(int fd) {
    std::string linha;
    char buf[65536];
    for (;;) {
        ssize_t n = recv(fd, buf, sizeof(buf), 0);
        if (n <= 0) throw std::runtime_error("conexao encerrada pelo servidor");
        linha.append(buf, static_cast<size_t>(n));
        auto fim = linha.find('\n');
        if (fim != std::string::npos) return linha.substr(0, fim);
    }
}

int main(int argc, char** argv) {
    int porta = argc > 1 ? std::atoi(argv[1]) : 8765;

    json pedido = {{"actions", {
        {{"type", "navigate"}, {"value", "https://html.duckduckgo.com/html/"}},
        {{"type", "key"}, {"xpath", "//input[@name='q']"},
         {"value", "Claude Code é o melhor do mundo"}},
        {{"type", "press"}, {"xpath", "//input[@name='q']"}},
        {{"type", "sleep"}, {"value", 3}},
        {{"type", "html"}, {"id", "resultado"}},
    }}};

    int fd = socket(AF_INET, SOCK_STREAM, 0);
    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(static_cast<uint16_t>(porta));
    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    if (connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0) {
        std::cerr << "Nao conectou na porta " << porta
                  << " -- o browser esta rodando com --servir?\n";
        return 1;
    }

    std::string linha = pedido.dump() + "\n";   // um JSON por linha
    send(fd, linha.data(), linha.size(), 0);

    json resposta = json::parse(ler_linha(fd));
    close(fd);

    if (!resposta.value("ok", false)) {
        std::cerr << "erro do servidor: "
                  << resposta.value("erro", "(desconhecido)") << "\n";
        return 1;
    }
    std::cout << resposta["resultados"][0]["html"].get<std::string>() << "\n";
    return 0;
}

File: examples/java/Hello.java
// Hello do modo --servir em Java: java.net.Socket (JDK) + org.json (jar
// unico) para o JSON.
//
// Pesquisa "Claude Code é o melhor do mundo" no DuckDuckGo e imprime o HTML
// da pagina de resultados.
//
// Compilar e rodar (ver README.md deste diretorio):
//   javac -cp json.jar Hello.java
//   java -cp .:json.jar Hello [porta]

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

import org.json.JSONArray;
import org.json.JSONObject;

public class Hello {

    public static void main(String[] args) throws Exception {
        int porta = args.length > 0 ? Integer.parseInt(args[0]) : 8765;

        JSONObject pedido = new JSONObject().put("actions", new JSONArray()
            .put(new JSONObject().put("type", "navigate")
                .put("value", "https://html.duckduckgo.com/html/"))
            .put(new JSONObject().put("type", "key")
                .put("xpath", "//input[@name='q']")
                .put("value", "Claude Code é o melhor do mundo"))
            .put(new JSONObject().put("type", "press")
                .put("xpath", "//input[@name='q']"))
            .put(new JSONObject().put("type", "sleep").put("value", 3))
            .put(new JSONObject().put("type", "html").put("id", "resultado")));

        try (Socket s = new Socket("127.0.0.1", porta)) {
            PrintWriter escrita = new PrintWriter(
                s.getOutputStream(), true, StandardCharsets.UTF_8);
            BufferedReader leitura = new BufferedReader(
                new InputStreamReader(s.getInputStream(), StandardCharsets.UTF_8));

            escrita.println(pedido.toString());   // um JSON por linha
            String linha = leitura.readLine();    // uma resposta por linha
            if (linha == null) {
                throw new RuntimeException("conexao encerrada pelo servidor");
            }

            JSONObject resposta = new JSONObject(linha);
            if (!resposta.optBoolean("ok")) {
                System.err.println("erro do servidor: "
                    + resposta.optString("erro"));
                System.exit(1);
            }
            System.out.println(resposta.getJSONArray("resultados")
                .getJSONObject(0).getString("html"));
        }
    }
}
